File: MiniDOMFinanziaria/Finanziaria/Pratiche/CInfoStato.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica



Partial Public Class Finanziaria

    ''' <summary>
    ''' Passaggio di stato usato nel cursore delle pratiche
    ''' </summary>
    <Serializable>
    Public Class CInfoStato
        Inherits CIntervalloData

        Public IDOperatore As Integer
        Public NomeOperatore As String
        Public Parameters As String
        Public Note As String
        Public Forzato As Boolean?
        Public IDFromStato As Integer
        Public IDToStato As Integer
        Public IDOfferta As Integer
        Public IDRegolaApplicata As Integer
        Public NomeRegolaApplicata As String
        Public Descrizione As String
        Public MacroStato As StatoPraticaEnum?

        ''' <summary>
        ''' Costruttore
        ''' </summary>
        Public Sub New()
            Me.IDOperatore = 0
            Me.NomeOperatore = ""
            Me.Parameters = ""
            Me.Note = ""
            Me.Forzato = Nothing
            Me.IDFromStato = 0
            Me.IDToStato = 0
            Me.IDOfferta = 0
            Me.IDRegolaApplicata = 0
            Me.NomeRegolaApplicata = ""
            Me.Descrizione = ""
            Me.MacroStato = Nothing
        End Sub

        ''' <summary>
        ''' Costruttore
        ''' </summary>
        ''' <param name="ms"></param>
        Public Sub New(ByVal ms As StatoPraticaEnum)
            Me.New
            Me.MacroStato = ms
        End Sub

        Public Overrides Function IsSet() As Boolean
            Dim intervallo As CIntervalloData = DateUtils.PeriodoToDates(Me.Tipo, Me.Inizio, Me.Fine)
            Me.Inizio = intervallo.Inizio
            Me.Fine = intervallo.Fine
            Return (Me.Inizio.HasValue) OrElse (Me.Fine.HasValue) OrElse
                (Me.IDOperatore <> 0) OrElse (Me.NomeOperatore <> "") OrElse
                (Me.Parameters <> "") OrElse (Me.Note <> "") OrElse (Me.Forzato.HasValue) OrElse
                (Me.IDFromStato <> 0) OrElse (Me.IDOfferta <> 0) OrElse
                (Me.IDRegolaApplicata <> 0) OrElse (Me.NomeRegolaApplicata <> "") OrElse
                (Me.Descrizione <> "") OrElse (Me.IDToStato <> 0) 'OrElse (Me.MacroStato.HasValue)
        End Function

        Public Overrides Sub Clear()
            MyBase.Clear()
            Me.IDOperatore = 0
            Me.NomeOperatore = ""
            Me.Parameters = ""
            Me.Note = ""
            Me.Forzato = Nothing
            Me.IDFromStato = 0
            Me.IDOfferta = 0
            Me.IDRegolaApplicata = 0
            Me.NomeRegolaApplicata = ""
            Me.Descrizione = ""
            Me.IDToStato = 0
            'Me.MacroStato = Nothing
        End Sub

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("IDOperatore", Me.IDOperatore)
            writer.WriteAttribute("NomeOperatore", Me.NomeOperatore)
            writer.WriteAttribute("Forzato", Me.Forzato)
            writer.WriteAttribute("IDFromStato", Me.IDFromStato)
            writer.WriteAttribute("IDToStato", Me.IDToStato)
            writer.WriteAttribute("IDOfferta", Me.IDOfferta)
            writer.WriteAttribute("IDRegolaApplicata", Me.IDRegolaApplicata)
            writer.WriteAttribute("NomeRegolaApplicata", Me.NomeRegolaApplicata)
            writer.WriteAttribute("Descrizione", Me.Descrizione)
            writer.WriteAttribute("MacroStato", Me.MacroStato)
            writer.WriteAttribute("Parameters", Me.Parameters)
            MyBase.XMLSerialize(writer)
            writer.WriteTag("Note", Me.Note)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "IDOperatore" : Me.IDOperatore = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "NomeOperatore" : Me.NomeOperatore = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Forzato" : Me.Forzato = XML.Utils.Serializer.DeserializeBoolean(fieldValue)
                Case "IDFromStato" : Me.IDFromStato = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "IDToStato" : Me.IDToStato = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "IDOfferta" : Me.IDOfferta = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "IDRegolaApplicata" : Me.IDRegolaApplicata = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "NomeRegolaApplicata" : Me.NomeRegolaApplicata = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Descrizione" : Me.Descrizione = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Note" : Me.Note = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Parameters" : Me.Parameters = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "MacroStato" : Me.MacroStato = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select
        End Sub

        Public Overrides Sub CopyFrom(value As Object)
            With DirectCast(value, CInfoStato)
                Me.IDOperatore = .IDOperatore
                Me.NomeOperatore = .NomeOperatore
                Me.Parameters = .Parameters
                Me.Note = .Note
                Me.Forzato = .Forzato
                Me.IDFromStato = .IDFromStato
                Me.IDOfferta = .IDOfferta
                Me.IDRegolaApplicata = .IDRegolaApplicata
                Me.NomeRegolaApplicata = .NomeRegolaApplicata
                Me.Descrizione = .Descrizione
                Me.IDToStato = .IDToStato
            End With
            MyBase.CopyFrom(value)
        End Sub

        Public Function GetSQL() As String
            Dim intervallo As CIntervalloData = DateUtils.PeriodoToDates(Me.Tipo, Me.Inizio, Me.Fine)
            Me.Inizio = intervallo.Inizio
            Me.Fine = intervallo.Fine
            
            Dim wherePart As String = ""

            If (Me.IDOperatore <> 0) Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDOperatore]=" & DBUtils.DBNumber(Me.IDOperatore), " AND ")
            If (Me.NomeOperatore <> "") Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[NomeOperatore]=" & DBUtils.DBString(Me.NomeOperatore), " AND ")
            If (Me.Parameters <> "") Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[Parameters]=" & DBUtils.DBString(Me.Parameters), " AND ")
            If (Me.Note <> "") Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[Note] Like '" & Strings.Replace(Me.Note, "'", "''") & "%')", " And ")
            If (Me.Forzato.HasValue) Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[Forzato]=" & DBUtils.DBBool(Me.Forzato.Value), " And ")
            If (Me.IDFromStato <> 0) Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDFromStato]=" & DBUtils.DBNumber(Me.IDFromStato), " AND ")
            If (Me.IDToStato <> 0) Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDToStato]=" & DBUtils.DBNumber(Me.IDToStato), " AND ")
            If (Me.IDOfferta <> 0) Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDOfferta]=" & DBUtils.DBNumber(Me.IDOfferta), " AND ")
            If (Me.IDRegolaApplicata <> 0) Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDRegolaApplicata]=" & DBUtils.DBNumber(Me.IDRegolaApplicata), " AND ")
            If (Me.NomeRegolaApplicata <> "") Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[NomeRegolaApplicata]=" & DBUtils.DBString(Me.NomeRegolaApplicata), " AND ")
            If (Me.Descrizione <> "") Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[Descrizione]=" & DBUtils.DBString(Me.Descrizione), " AND ")
            If (Me.MacroStato.HasValue) Then wherePart = Strings.Combine(wherePart, "[tbl_PraticheSTL].[MacroStato]=" & DBUtils.DBNumber(Me.MacroStato.Value), " AND ")
            If (Me.Fine.HasValue) Then Me.Fine = DateUtils.DateAdd(DateInterval.Second, 3600 * 24 - 1, DateUtils.GetDatePart(Me.Fine.Value))
            If (Me.Inizio.HasValue) Then wherePart = Strings.Combine(wherePart, "([tbl_PraticheSTL].[Data] >= " & DBUtils.DBDate(Me.Inizio.Value) & ")", " AND ")
            If (Me.Fine.HasValue) Then wherePart = Strings.Combine(wherePart, "([tbl_PraticheSTL].[Data] <= " & DBUtils.DBDate(Me.Fine.Value) & ")", " AND ")

            Return wherePart
        End Function

        Public Function GetCampoSQL( ) As String
            Dim intervallo As CIntervalloData = DateUtils.PeriodoToDates(Me.Tipo, Me.Inizio, Me.Fine)
            Me.Inizio = intervallo.Inizio
            Me.Fine = intervallo.Fine
            If Me.IsSet = False Then Return ""

            Dim wherePart As String = "[Stato]=" & ObjectStatus.OBJECT_VALID
            If (Me.IDOperatore <> 0) Then wherePart = Strings.Combine(wherePart, "[IDOperatore]=" & DBUtils.DBNumber(Me.IDOperatore), " AND ")
            If (Me.NomeOperatore <> "") Then wherePart = Strings.Combine(wherePart, "[NomeOperatore]=" & DBUtils.DBString(Me.NomeOperatore), " AND ")
            If (Me.Parameters <> "") Then wherePart = Strings.Combine(wherePart, "[Parameters]=" & DBUtils.DBString(Me.Parameters), " AND ")
            If (Me.Note <> "") Then wherePart = Strings.Combine(wherePart, "[Note] Like '%" & Strings.Replace(Me.Note, "'", "''") & "%')", " And ")
            If (Me.Forzato.HasValue) Then wherePart = Strings.Combine(wherePart, "[Forzato]=" & DBUtils.DBBool(Me.Forzato.Value), " And ")
            If (Me.IDFromStato <> 0) Then wherePart = Strings.Combine(wherePart, "[IDFromStato]=" & DBUtils.DBNumber(Me.IDFromStato), " AND ")
            If (Me.IDToStato <> 0) Then wherePart = Strings.Combine(wherePart, "[IDToStato]=" & DBUtils.DBNumber(Me.IDToStato), " AND ")
            If (Me.IDOfferta <> 0) Then wherePart = Strings.Combine(wherePart, "[IDOfferta]=" & DBUtils.DBNumber(Me.IDOfferta), " AND ")
            If (Me.IDRegolaApplicata <> 0) Then wherePart = Strings.Combine(wherePart, "[IDRegolaApplicata]=" & DBUtils.DBNumber(Me.IDRegolaApplicata), " AND ")
            If (Me.NomeRegolaApplicata <> "") Then wherePart = Strings.Combine(wherePart, "[NomeRegolaApplicata]=" & DBUtils.DBString(Me.NomeRegolaApplicata), " AND ")
            If (Me.Descrizione <> "") Then wherePart = Strings.Combine(wherePart, "[Descrizione]=" & DBUtils.DBString(Me.Descrizione), " AND ")
            If (Me.MacroStato.HasValue) Then wherePart = Strings.Combine(wherePart, "[MacroStato]=" & DBUtils.DBNumber(Me.MacroStato.Value), " AND ")
            If (Me.Fine.HasValue) Then Me.Fine = DateUtils.DateAdd(DateInterval.Second, 3600 * 24 - 1, DateUtils.GetDatePart(Me.Fine.Value))
            If (Me.Inizio.HasValue) Then wherePart = Strings.Combine(wherePart, "([Data] >= " & DBUtils.DBDate(Me.Inizio.Value) & ")", " AND ")
            If (Me.Fine.HasValue) Then wherePart = Strings.Combine(wherePart, "([Data] <= " & DBUtils.DBDate(Me.Fine.Value) & ")", " AND ")
            Return "([ID] In (SELECT [IDPratica] FROM [tbl_PraticheSTL] WHERE " & wherePart & "))"
        End Function


    End Class

End Class

File: MiniDOMFinanziaria/Finanziaria/Pratiche/CInfoStato.cs
using System;
using DMD;
using DMD.XML;
using minidom;
using static minidom.Sistema;

namespace minidom
{
    public partial class Finanziaria
    {

        /// <summary>
        /// Passaggio di stato usato nel cursore delle pratiche
        /// </summary>
        [Serializable]
        public class CInfoStato 
            : CIntervalloData
        {
            public int IDOperatore;
            public string NomeOperatore;
            public string Parameters;
            public string Note;
            public bool? Forzato;
            public int IDFromStato;
            public int IDToStato;
            public int IDOfferta;
            public int IDRegolaApplicata;
            public string NomeRegolaApplicata;
            public string Descrizione;
            public StatoPraticaEnum? MacroStato;

            /// <summary>
        /// Costruttore
        /// </summary>
            public CInfoStato()
            {
                IDOperatore = 0;
                NomeOperatore = "";
                Parameters = "";
                Note = "";
                Forzato = default;
                IDFromStato = 0;
                IDToStato = 0;
                IDOfferta = 0;
                IDRegolaApplicata = 0;
                NomeRegolaApplicata = "";
                Descrizione = "";
                MacroStato = default;
            }

            /// <summary>
        /// Costruttore
        /// </summary>
        /// <param name="ms"></param>
            public CInfoStato(StatoPraticaEnum ms) : this()
            {
                MacroStato = ms;
            }

            public override bool IsSet()
            {
                var intervallo = DMD.DateUtils.PeriodoToDates(Tipo, Inizio, Fine);
                Inizio = intervallo.Inizio;
                Fine = intervallo.Fine;
                return Inizio.HasValue || Fine.HasValue || IDOperatore != 0 || !string.IsNullOrEmpty(NomeOperatore) || !string.IsNullOrEmpty(Parameters) || !string.IsNullOrEmpty(Note) || Forzato.HasValue || IDFromStato != 0 || IDOfferta != 0 || IDRegolaApplicata != 0 || !string.IsNullOrEmpty(NomeRegolaApplicata) || !string.IsNullOrEmpty(Descrizione) || IDToStato != 0; // OrElse (Me.MacroStato.HasValue)
            }

            public override void Clear()
            {
                base.Clear();
                IDOperatore = 0;
                NomeOperatore = "";
                Parameters = "";
                Note = "";
                Forzato = default;
                IDFromStato = 0;
                IDOfferta = 0;
                IDRegolaApplicata = 0;
                NomeRegolaApplicata = "";
                Descrizione = "";
                IDToStato = 0;
                // Me.MacroStato = Nothing
            }

            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("IDOperatore", IDOperatore);
                writer.WriteAttribute("NomeOperatore", NomeOperatore);
                writer.WriteAttribute("Forzato", Forzato);
                writer.WriteAttribute("IDFromStato", IDFromStato);
                writer.WriteAttribute("IDToStato", IDToStato);
                writer.WriteAttribute("IDOfferta", IDOfferta);
                writer.WriteAttribute("IDRegolaApplicata", IDRegolaApplicata);
                writer.WriteAttribute("NomeRegolaApplicata", NomeRegolaApplicata);
                writer.WriteAttribute("Descrizione", Descrizione);
                writer.WriteAttribute("MacroStato", MacroStato);
                writer.WriteAttribute("Parameters", Parameters);
                base.XMLSerialize(writer);
                writer.WriteTag("Note", Note);
            }

            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "IDOperatore":
                        {
                            IDOperatore = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "NomeOperatore":
                        {
                            NomeOperatore = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Forzato":
                        {
                            Forzato = DMD.XML.Utils.Serializer.DeserializeBoolean(fieldValue);
                            break;
                        }

                    case "IDFromStato":
                        {
                            IDFromStato = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "IDToStato":
                        {
                            IDToStato = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "IDOfferta":
                        {
                            IDOfferta = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "IDRegolaApplicata":
                        {
                            IDRegolaApplicata = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "NomeRegolaApplicata":
                        {
                            NomeRegolaApplicata = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Descrizione":
                        {
                            Descrizione = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Note":
                        {
                            Note = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Parameters":
                        {
                            Parameters = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "MacroStato":
                        {
                            MacroStato = (StatoPraticaEnum?)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

            public override void InitializeFrom(object value)
            {
                {
                    var withBlock = (CInfoStato)value;
                    IDOperatore = withBlock.IDOperatore;
                    NomeOperatore = withBlock.NomeOperatore;
                    Parameters = withBlock.Parameters;
                    Note = withBlock.Note;
                    Forzato = withBlock.Forzato;
                    IDFromStato = withBlock.IDFromStato;
                    IDOfferta = withBlock.IDOfferta;
                    IDRegolaApplicata = withBlock.IDRegolaApplicata;
                    NomeRegolaApplicata = withBlock.NomeRegolaApplicata;
                    Descrizione = withBlock.Descrizione;
                    IDToStato = withBlock.IDToStato;
                }

                base.InitializeFrom(value);
            }

            protected override void _CopyFrom(object value)
            {
                var withBlock = (CInfoStato)value;
                IDOperatore = withBlock.IDOperatore;
                NomeOperatore = withBlock.NomeOperatore;
                Parameters = withBlock.Parameters;
                Note = withBlock.Note;
                Forzato = withBlock.Forzato;
                IDFromStato = withBlock.IDFromStato;
                IDOfferta = withBlock.IDOfferta;
                IDRegolaApplicata = withBlock.IDRegolaApplicata;
                NomeRegolaApplicata = withBlock.NomeRegolaApplicata;
                Descrizione = withBlock.Descrizione;
                IDToStato = withBlock.IDToStato;
                base._CopyFrom(value);
            }

            public string GetSQL()
            {
                var intervallo = DMD.DateUtils.PeriodoToDates(Tipo, Inizio, Fine);
                Inizio = intervallo.Inizio;
                Fine = intervallo.Fine;
                string wherePart = "";
                if (IDOperatore != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDOperatore]=" + DBUtils.DBNumber(IDOperatore), " AND ");
                if (!string.IsNullOrEmpty(NomeOperatore))
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[NomeOperatore]=" + DBUtils.DBString(NomeOperatore), " AND ");
                if (!string.IsNullOrEmpty(Parameters))
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[Parameters]=" + DBUtils.DBString(Parameters), " AND ");
                if (!string.IsNullOrEmpty(Note))
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[Note] Like '" + DMD.Strings.Replace(Note, "'", "''") + "%')", " And ");
                if (Forzato.HasValue)
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[Forzato]=" + DBUtils.DBBool(Forzato.Value), " And ");
                if (IDFromStato != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDFromStato]=" + DBUtils.DBNumber(IDFromStato), " AND ");
                if (IDToStato != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDToStato]=" + DBUtils.DBNumber(IDToStato), " AND ");
                if (IDOfferta != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDOfferta]=" + DBUtils.DBNumber(IDOfferta), " AND ");
                if (IDRegolaApplicata != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[IDRegolaApplicata]=" + DBUtils.DBNumber(IDRegolaApplicata), " AND ");
                if (!string.IsNullOrEmpty(NomeRegolaApplicata))
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[NomeRegolaApplicata]=" + DBUtils.DBString(NomeRegolaApplicata), " AND ");
                if (!string.IsNullOrEmpty(Descrizione))
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[Descrizione]=" + DBUtils.DBString(Descrizione), " AND ");
                if (MacroStato.HasValue)
                    wherePart = DMD.Strings.Combine(wherePart, "[tbl_PraticheSTL].[MacroStato]=" + DBUtils.DBNumber(MacroStato.Value), " AND ");
                if (Fine.HasValue)
                    Fine = DMD.DateUtils.DateAdd(DateTimeInterval.Second, 3600 * 24 - 1, DMD.DateUtils.GetDatePart(Fine.Value));
                if (Inizio.HasValue)
                    wherePart = DMD.Strings.Combine(wherePart, "([tbl_PraticheSTL].[Data] >= " + DBUtils.DBDate(Inizio.Value) + ")", " AND ");
                if (Fine.HasValue)
                    wherePart = DMD.Strings.Combine(wherePart, "([tbl_PraticheSTL].[Data] <= " + DBUtils.DBDate(Fine.Value) + ")", " AND ");
                return wherePart;
            }

            public string GetCampoSQL()
            {
                var intervallo = DMD.DateUtils.PeriodoToDates(Tipo, Inizio, Fine);
                Inizio = intervallo.Inizio;
                Fine = intervallo.Fine;
                if (IsSet() == false)
                    return "";
                string wherePart = "[Stato]=" + ((int)ObjectStatus.OBJECT_VALID).ToString();
                if (IDOperatore != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[IDOperatore]=" + DBUtils.DBNumber(IDOperatore), " AND ");
                if (!string.IsNullOrEmpty(NomeOperatore))
                    wherePart = DMD.Strings.Combine(wherePart, "[NomeOperatore]=" + DBUtils.DBString(NomeOperatore), " AND ");
                if (!string.IsNullOrEmpty(Parameters))
                    wherePart = DMD.Strings.Combine(wherePart, "[Parameters]=" + DBUtils.DBString(Parameters), " AND ");
                if (!string.IsNullOrEmpty(Note))
                    wherePart = DMD.Strings.Combine(wherePart, "[Note] Like '%" + DMD.Strings.Replace(Note, "'", "''") + "%')", " And ");
                if (Forzato.HasValue)
                    wherePart = DMD.Strings.Combine(wherePart, "[Forzato]=" + DBUtils.DBBool(Forzato.Value), " And ");
                if (IDFromStato != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[IDFromStato]=" + DBUtils.DBNumber(IDFromStato), " AND ");
                if (IDToStato != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[IDToStato]=" + DBUtils.DBNumber(IDToStato), " AND ");
                if (IDOfferta != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[IDOfferta]=" + DBUtils.DBNumber(IDOfferta), " AND ");
                if (IDRegolaApplicata != 0)
                    wherePart = DMD.Strings.Combine(wherePart, "[IDRegolaApplicata]=" + DBUtils.DBNumber(IDRegolaApplicata), " AND ");
                if (!string.IsNullOrEmpty(NomeRegolaApplicata))
                    wherePart = DMD.Strings.Combine(wherePart, "[NomeRegolaApplicata]=" + DBUtils.DBString(NomeRegolaApplicata), " AND ");
                if (!string.IsNullOrEmpty(Descrizione))
                    wherePart = DMD.Strings.Combine(wherePart, "[Descrizione]=" + DBUtils.DBString(Descrizione), " AND ");
                if (MacroStato.HasValue)
                    wherePart = DMD.Strings.Combine(wherePart, "[MacroStato]=" + DBUtils.DBNumber(MacroStato.Value), " AND ");
                if (Fine.HasValue)
                    Fine = DMD.DateUtils.DateAdd(DateTimeInterval.Second, 3600 * 24 - 1, DMD.DateUtils.GetDatePart(Fine.Value));
                if (Inizio.HasValue)
                    wherePart = DMD.Strings.Combine(wherePart, "([Data] >= " + DBUtils.DBDate(Inizio.Value) + ")", " AND ");
                if (Fine.HasValue)
                    wherePart = DMD.Strings.Combine(wherePart, "([Data] <= " + DBUtils.DBDate(Fine.Value) + ")", " AND ");
                return "([ID] In (SELECT [IDPratica] FROM [tbl_PraticheSTL] WHERE " + wherePart + "))";
            }
        }
    }
}